### ISSUE TITLE
Segmentation fault with x86 "loop" instruction

### ISSUE THREAD

**Issue by niooss-ledger** [QBDI/QBDI#199]
**Bug description**

Hello, I am trying to use PyQBDI to study a piece of x86-32 code. In this work, I encountered a segmentation fault. I managed to reproduce this issue in the following reproducer program, which works for both x86-32 and x86-64:

```python
import ctypes
import ctypes.util
import pyqbdi


"""
Simple function which loops some nop instructions and returns 1:

In x86-32:
   0:	51                   	push   ecx
   1:	b9 02 00 00 00       	mov    ecx,0x2
   6:	90                   	nop
   7:	90                   	nop
   8:	e2 fc                	loop   0x6
   a:	59                   	pop    ecx
   b:	b8 01 00 00 00       	mov    eax,0x1
  10:	c3                   	ret

In x86-64:
   0:	51                   	push   rcx
   1:	b9 02 00 00 00       	mov    ecx,0x2
   6:	90                   	nop
   7:	90                   	nop
   8:	e2 fc                	loop   0x6
   a:	59                   	pop    rcx
   b:	b8 01 00 00 00       	mov    eax,0x1
  10:	c3                   	ret
"""
loop_function = bytes.fromhex("51b9020000009090e2fc59b801000000c3")

# Allocate a RWX page and copy the loop function in it
libc = ctypes.CDLL(ctypes.util.find_library('c'))
libc.mmap.restype = ctypes.c_void_p
PROT_READ = 1
PROT_WRITE = 2
PROT_EXEC = 4
MAP_PRIVATE = 2
MAP_FIXED = 0x10
MAP_ANONYMOUS = 0x20

func_mem = libc.mmap(0, 0x1000, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)
assert (func_mem & 0xffffffff) != 0xffffffff
print(f"Using RWX page at {func_mem:#010x}")
ctypes.memmove(func_mem, loop_function, len(loop_function))

# Initialize QBDI VM
vm = pyqbdi.VM()
state = vm.getGPRState()
fakestack = pyqbdi.allocateVirtualStack(state, 0x100000)
assert fakestack
print(f"Fake stack allocated at {fakestack:#010x}")


# Trace the instructions and the values of ecx/rcx
def show_rcx(vm, gpr, fpr, data):
    instAnalysis = vm.getInstAnalysis()
    if ctypes.sizeof(ctypes.c_void_p) == 8:
        print(f"{instAnalysis.address:#010x}: rcx was {gpr.rcx:#x}")
    else:
        print(f"{instAnalysis.address:#010x}: ecx was {gpr.ecx:#x}")
    return pyqbdi.CONTINUE


cid = vm.addCodeCB(pyqbdi.PREINST, show_rcx, None)
assert cid != pyqbdi.INVALID_EVENTID


def show_instruction(vm, gpr, fpr, data):
    instAnalysis = vm.getInstAnalysis()
    print(f"{instAnalysis.address:#010x}: {instAnalysis.disassembly}")
    return pyqbdi.CONTINUE


cid = vm.addCodeCB(pyqbdi.PREINST, show_instruction, None)
assert cid != pyqbdi.INVALID_EVENTID

# Call the function, instrumenting it
vm.addInstrumentedRange(func_mem, func_mem + len(loop_function))
success, retval = vm.call(func_mem, [])
assert success
print(f"Return value is {retval}")
pyqbdi.alignedFree(fakestack)
```

On x86-64, this leads to:

```text
Using RWX page at 0x7fbf43c75000
Fake stack allocated at 0x7fbf425a3010
0x7fbf43c75000: rcx was 0x0
0x7fbf43c75000: 	push	rcx
0x7fbf43c75001: rcx was 0x0
0x7fbf43c75001: 	mov	ecx, 2
0x7fbf43c75006: rcx was 0x2
0x7fbf43c75006: 	nop
0x7fbf43c75007: rcx was 0x2
0x7fbf43c75007: 	nop
0x7fbf43c75008: rcx was 0x2
0x7fbf43c75008: 	loop	-4
Segmentation fault (core dumped)
```

On x86-32, this leads to:

```text
Using RWX page at 0xf7f3b000
Fake stack allocated at 0xf6e3c010
0xf7f3b000: 	push	ecx
0xf7f3b000: ecx was 0x0
0xf7f3b001: 	mov	ecx, 2
0xf7f3b001: ecx was 0x0
0xf7f3b006: 	nop
0xf7f3b006: ecx was 0x2
0xf7f3b007: 	nop
0xf7f3b007: ecx was 0x2
0xf7f3b008: 	loop	-4
0xf7f3b008: ecx was 0x2
Segmentation fault (core dumped)
```

**Bug log**
If I replace `mov    ecx,0x2` with `mov    ecx,0x1` in the emulated assembly, the emulation works correctly:

```text
Using RWX page at 0x7fabd5c82000
Fake stack allocated at 0x7fabd45b0010
0x7fabd5c82000: rcx was 0x0
0x7fabd5c82000: 	push	rcx
0x7fabd5c82001: rcx was 0x0
0x7fabd5c82001: 	mov	ecx, 1
0x7fabd5c82006: rcx was 0x1
0x7fabd5c82006: 	nop
0x7fabd5c82007: rcx was 0x1
0x7fabd5c82007: 	nop
0x7fabd5c82008: rcx was 0x1
0x7fabd5c82008: 	loop	-4
0x7fabd5c8200a: rcx was 0x0
0x7fabd5c8200a: 	pop	rcx
0x7fabd5c8200b: rcx was 0x0
0x7fabd5c8200b: 	mov	eax, 1
0x7fabd5c82010: rcx was 0x0
0x7fabd5c82010: 	ret
Return value is 1
```

So the issue really occurs when the `loop -4` instruction branches back to the first `nop` instruction.

**Minimal code to reproduce the bug**
See above

**Specific command to reproduce the bug**

I used a podman container to reproduce against the current `dev-next` branch:

```sh
podman run --rm -ti docker.io/library/python:3.9-slim bash

apt-get update && apt-get install -y git build-essential cmake g++ ninja-build
git clone https://github.com/QBDI/QBDI
mkdir QBDI/build
cd QBDI/build
../cmake/config/config-linux-X86_64.sh
ninja
export PYTHONPATH="$(pwd)/tools/pyqbdi"
python3 /reproducer.py
```

**Version (please complete the following information):**
- OS: Linux
- Architecture: X86 and X64
- QBDI version (returned by getVersion) and commit hash (if not a release version): both latest PyPI package (`pyqbdi.__version__ = '0.8.0'`) and the latest `dev-next` branch (`pyqbdi.__version__ = '0.8.1-devel-4b9f3e0'`, commit 4b9f3e07dbed60456d3198fa733af06d9da2e558)
- Frida version (for frida/QBDI issue): not applicable
- Python version (for PyQBDI issue): Python 3.9.7 (reproduced using container image `docker.io/library/python:3.9-slim`)

**Additional context**
I am working with a piece of code which comes from a raw file (not a shared library), which is why I called `mmap`. Is there is a way which is more suitable to load raw code?

Moreover in this reproducer, I called `vm.addCodeCB` twice. If I merge them both, the program no longer fails on x86-32, but the execution then seems to be incorrect (the `loop` instruction seems to never branch).

**Comment by nsurbay** [comment]
Hello,

I have never see the instruction `loop`. The issue is likely in the core of QBDI that doesn't have a PatchRule for this instruction (see [PatchRules](https://github.com/QBDI/QBDI/blob/dev-next/src/Patch/X86_64/PatchRules_X86_64.cpp#L449)). The instruction is not patch and jump in the middle of the instrumented code.

I will try to provide a patch for this instruction quickly.

**Comment by nsurbay** [comment]
I push a fix in the MR #200. With it, the result should be the expected one:
```
Using RWX page at 0x7f99d1aaa000
Fake stack allocated at 0x7f99d00d0010
0x7f99d1aaa000: rcx was 0x0
0x7f99d1aaa000: 	push	rcx
0x7f99d1aaa001: rcx was 0x0
0x7f99d1aaa001: 	mov	ecx, 2
0x7f99d1aaa006: rcx was 0x2
0x7f99d1aaa006: 	nop
0x7f99d1aaa007: rcx was 0x2
0x7f99d1aaa007: 	nop
0x7f99d1aaa008: rcx was 0x2
0x7f99d1aaa008: 	loop	-4
0x7f99d1aaa006: rcx was 0x1
0x7f99d1aaa006: 	nop
0x7f99d1aaa007: rcx was 0x1
0x7f99d1aaa007: 	nop
0x7f99d1aaa008: rcx was 0x1
0x7f99d1aaa008: 	loop	-4
0x7f99d1aaa00a: rcx was 0x0
0x7f99d1aaa00a: 	pop	rcx
0x7f99d1aaa00b: rcx was 0x0
0x7f99d1aaa00b: 	mov	eax, 1
0x7f99d1aaa010: rcx was 0x0
0x7f99d1aaa010: 	ret
Return value is 1
```
> Is there is a way which is more suitable to load raw code?

An mmap of `rwx` region is the easier way to understand. However, QBDI will not execute the code. It will read it, create an instrumented assembly and execute this assembly. The original code can be allocate in rw- on the heap and QBDI should work without issue.

```python
loop_function = bytes.fromhex("51b9020000009090e2fc59b801000000c3")

# allocate a buffer on the heap
loop_raw_function = pyqbdi.allocateMemory(len(loop_function))
# copy the bytecode to the buffer
pyqbdi.writeMemory(loop_raw_function, loop_function)

# add the buffer in the InstrumentedRange
vm.addInstrumentedRange(loop_raw_function, loop_raw_function + len(loop_function))
# [optionnal] clear any previous cache in the VM.
# This should be done if you change the content of the buffer.
vm.clearCache(loop_raw_function, loop_raw_function + len(loop_function))
```

**Comment by niooss-ledger** [comment]
Thanks for your quick fix! I confirm it fixes the issue I had :)
